core/utils/codel.h

```diff
@@ -0,0 +1,267 @@
+#ifndef BESS_UTILS_CODEL_H_
+#define BESS_UTILS_CODEL_H_
+
+#include <math.h>
+#include <sys/time.h>
+
+#include <iostream>
+
+#include "./lock_free_queue.h"
+#include "./queue.h"
+
+namespace bess {
+namespace utils {
+// Codel(Controlled Delay Management) is an Queue controller based on this
+// article http://queue.acm.org/detail.cfm?id=2209336
+
+// It provides an active queue management to help prevent bufferbloat by dropping
+// queue entries at an increasing rate if the delay in the queue is above the
+// target queue delay. The equation used to calculate drop intervals is based on TCP
+// throughput response to drop probability.
+
+// Takes Template argument Q which is a templated child class of Queue and T the type
+// that is going to be enqueued/dequeued.
+template <typename T>
+class Codel : public Queue<T> {
+ public:
+  static constexpr uint64_t kDefaultTarget =
+      5000;  // default delay target for codel
+  static constexpr uint64_t kDefaultWindow =
+      100000;  // default window size for codel
+  static const int kDefaultSlots =
+      4096;  // default number of slots in the codel queue
+
+  // A wrapper class around the pointer to track the delay time and whether
+  // the packet can be dropped after dequeueing.
+  struct Wrapper {
+    T* ptr;              // a pointer to be enqueued
+    uint64_t timestamp;  // a timestamp of the time the pointer was enqueued
+    uint8_t
+        drop;  // used during dequeue to determin if the packet can be dropped
+    Wrapper(T* obj, uint64_t stamp) : ptr(obj), timestamp(stamp), drop(0) {}
+  };
+
+  // Takes a drop function which is a function that should take dropped value
+  // and handle it removing the object potentially including freeing the object,
+  // init queue should be a pointer to a initialized Queue object, target is the target
+  // delay in microseconds and the window is the buffer time in microsecond
+  // before changing into drop state.
+  Codel(void (*drop_func)(T*), Queue<Wrapper>* init_queue,
+        uint64_t target = kDefaultTarget, uint64_t window = kDefaultWindow)
+      : delay_target_(target),
+        window_(window),
+        time_above_target_(0),
+        time_till_drop_(MicroSecondTime() + window),
+        drop_count_(0),
+        dropping_(0),
+        queue_(init_queue),
+        drop_func_(drop_func) {}
+
+  virtual ~Codel() {
+    Wrapper* w;
+    while ((w = queue_->Pop())) {
+      Drop(w);
+    }
+    delete queue_;
+  }
+
+  // Enqueue one object. Takes a pointer to the object to be added. Returns 0 on
+  // success.
+  int Push(T* ptr) {
+    Wrapper* w = new Wrapper(ptr, MicroSecondTime());
+    int err = queue_->Push(w);
+    if (err) {
+      delete w;
+    }
+    return err;
+  }
+
+  // Enqueue multiple objects. Takes a pointer an table of object pointers, the
+  // number of objects to be added. Returns 0 on success.
+  int Push(T** ptr, size_t count) {
+    if (Capacity() - Size() <= count) {
+      return -2;
```

Comment:
llring has this interesting error scheme where -1 is when the enqueue is successful but the queue is now full and -2 is when the enqueue is unsuccessful because there is no room. Should I extend this support into Codel? if not, should I change this in the llring wrapper as well?